scipy/stats/_covariance.py

```diff
@@ -0,0 +1,340 @@
+        A = np.atleast_2d(A)
+        if not np.issubdtype(A.dtype, np.number):
+            message = (f"The input `{name}` must be an array of numbers.")
+            raise ValueError(message)
+        m, n = A.shape[-2:]
+        if m != n:
+            message = (f"`{name}.shape[-2]` must equal `{name}.shape[-1]`")
+            raise ValueError(message)
+        return A
+
+    def _validate_vector(self, A, name):
+        A = np.atleast_1d(A)
+        if not np.issubdtype(A.dtype, np.number):
+            message = (f"The input `{name}` must be an array of numbers.")
+            raise ValueError(message)
+        return A
+
+
+class CovViaDiagonal(Covariance):
+    """
+    Representation of a covariance provided via the diagonal
+    """
+
+    def __init__(self, diagonal):
+        diagonal = self._validate_vector(diagonal, 'diagonal')
+
+        i_zero = diagonal <= 0
+        positive_diagonal = np.array(diagonal, dtype=np.float64)
+
+        positive_diagonal[i_zero] = 1  # ones don't affect determinant
+        self._log_pdet = np.sum(np.log(positive_diagonal), axis=-1)
+
+        psuedo_reciprocals = 1 / np.sqrt(positive_diagonal)
+        psuedo_reciprocals[i_zero] = 0
+
+        self._LP = psuedo_reciprocals
+        self._rank = positive_diagonal.shape[-1] - i_zero.sum(axis=-1)
+        self._A = np.apply_along_axis(np.diag, -1, diagonal)
+        self._dimensionality = diagonal.shape[-1]
+        self._i_zero = i_zero
+        self._shape = self._A.shape
+        self._allow_singular = True
+
+    def _whiten(self, x):
+        return _dot_diag(x, self._LP)
+
+    def _support_mask(self, x):
+        """
+        Check whether x lies in the support of the distribution.
+        """
+        return ~np.any(_dot_diag(x, self._i_zero), axis=-1)
+
+
+class CovViaPrecision(Covariance):
+    """
+    Representation of a covariance provided via the precision matrix
+    """
+
+    def __init__(self, precision, covariance=None):
+        precision = self._validate_matrix(precision, 'precision')
+        if covariance is not None:
+            covariance = self._validate_matrix(covariance, 'covariance')
+            message = "`precision.shape` must equal `covariance.shape`."
+            if precision.shape != covariance.shape:
+                raise ValueError(message)
+
+        self._LP = np.linalg.cholesky(precision)
+        self._log_pdet = -2*np.log(_extract_diag(self._LP)).sum(axis=-1)
+        self._rank = precision.shape[-1]  # must be full rank in invertible
+        self._precision = precision
+        self._covariance = covariance
+        self._dimensionality = self._rank
+        self._shape = precision.shape
+        self._allow_singular = False
+
+    def _whiten(self, x):
+        return x @ self._LP
+
+    @cached_property
+    def _A(self):
+        return (np.linalg.inv(self._precision) if self._covariance is None
+                else self._covariance)
+
+
+class CovViaEigendecomposition(Covariance):
```

Comment:
One advantage of the eigendecomposition is that it supports singular covariance matrices. But we can compute the key $x^T \Sigma^* x$ product using LDL factors, too. 

For nonsingular matrices, this is what it looks like:
```python3
import numpy as np
from scipy import linalg

n = 40
rng = np.random.default_rng(14)
x = rng.random(size=(n, 1))
A = rng.random(size=(n, n))
A = A @ A.T

w, V = linalg.eigh(A)
z = x.T @ (V * w**(-0.5))
xSxT = (z**2).sum()

L, D, perm = linalg.ldl(A)
y = linalg.solve_triangular(L[perm], x[perm], lower=True).T * np.diag(D)**(-0.5)
xSxT2 = (y**2).sum()

np.testing.assert_allclose(L, np.tril(L))

print(xSxT, xSxT2)
```

For singular matrices, we need to mask out the zero eigenvalues. When the vector `x` is within the right subspace, we can perform the calculation as:

```python3
import numpy as np
from scipy import linalg

n = 10
rng = np.random.default_rng(1329348965454623456)
x = rng.random(size=(n, 1))
A = rng.random(size=(n, n))

A[0] = A[1] + A[2]  # make it singular
x[0] = x[1] + x[2]
A = A @ A.T

eps = 1e-10

w, V = linalg.eigh(A)
mask = np.abs(w) < eps
w_ = w.copy()
w_[mask] = np.inf
z = x.T @ (V * w_**(-0.5))
xSxT = (z**2).sum()

L, D, perm = linalg.ldl(A)
d = np.diag(D).copy()[:, np.newaxis]
mask = np.abs(d) < eps
d_ = d.copy()
d[mask] = 0
d_[mask] = np.inf
y = linalg.solve_triangular(L[perm], x[perm], lower=True) * d_**(-0.5)
xSxT2 = (y**2).sum()

np.testing.assert_allclose(xSxT, xSxT2)
```

And we can tell it's in the subspace if `np.allclose(L@(d**0.5*y), x)`. Otherwise, the PDF is zero.

Reply:
This needs some work. The pseudodeterminant calculation isn't just the pseudodeterminant of the `D` matrix (product of positive diagonal elements).